### ISSUE TITLE
Allow parameterization of GCC_MULTILIBS?

### ISSUE THREAD

**Issue by TommyMurphyTM1234** [gnu-mcu-eclipse/riscv-none-gcc-build#10]
Small issue with the latest scripts - previous build scripts parameterized GCC_MULTILIBS so that a specific list of multilibs could be passed in to the scripts using the GCC_MULTILIBS env var. In the current container-build.sh script that is not possible.

Should the following lines:

https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/master/scripts/container-build.sh#L289
https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/master/scripts/container-build.sh#L350
https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/master/scripts/container-build.sh#L413

be changed from:

`GCC_MULTILIB=(rv32i-ilp32--c rv32im-ilp32 ...)}`

to

`GCC_MULTILIB=${GCC_MULTILIB:-(rv32i-ilp32--c ...)}`

**Comment by ilg-ul** [comment]
Oops!

You are right, `GCC_MULTILIB` is no longer configurable now.

Did you try to apply the suggested changes? If you confirm they are ok, I can apply them. Or you can submit a PR.

**Comment by ilg-ul** [comment]
Just curious, what other multilib you need? Can this be useful for the GNU MCU Eclipse distribution too?

**Comment by TommyMurphyTM1234** [comment]
Well we build multilibs for the "native" ABI of all architectures.
Basically this list of 39 multilibs:

```
rv32i-ilp32-- 
rv32ia-ilp32-- 
rv32iac-ilp32-- 
rv32iaf-ilp32f-- 
rv32iafc-ilp32f-- 
rv32ic-ilp32-- 
rv32if-ilp32f-- 
rv32ifc-ilp32f-- 
rv32im-ilp32-- 
rv32ima-ilp32-- 
rv32imac-ilp32-- 
rv32imaf-ilp32f-- 
rv32imafc-ilp32f-- 
rv32imc-ilp32-- 
rv32imf-ilp32f-- 
rv32imfc-ilp32f-- 
rv64i-lp64-- 
rv64ia-lp64-- 
rv64iac-lp64-- 
rv64iaf-lp64f-- 
rv64iafc-lp64f-- 
rv64iafd-lp64d-- 
rv64iafdc-lp64d-- 
rv64ic-lp64-- 
rv64if-lp64f-- 
rv64ifc-lp64f-- 
rv64ifd-lp64d-- 
rv64ifdc-lp64d-- 
rv64im-lp64-- 
rv64ima-lp64-- 
rv64imac-lp64-- 
rv64imaf-lp64f-- 
rv64imafc-lp64f-- 
rv64imafd-lp64d-- 
rv64imc-lp64-- 
rv64imf-lp64f-- 
rv64imfc-lp64f-- 
rv64imfd-lp64d-- 
rv64imfdc-lp64d--
```
But I'm not sure (a) what the total overhead is and (b) if everbody wants to carry that especially if they only use one or two?

**Comment by TommyMurphyTM1234** [comment]
I didn't do a PR yet - let me look at that.
Actually I made the change that I mentioned above but I'm not sure that the GCC_MULTILIB specified externally is being passed down to the docker from the host so for now I had to edit the script itself.

**Comment by ilg-ul** [comment]
> 39 multilibs ... what the total overhead is

hmmm... I think the overhead would be important, both in terms of build time and archive size.

how does your Windows archive size compare to mine?

**Comment by ilg-ul** [comment]
> I'm not sure that the GCC_MULTILIB specified externally is being passed down to the docker

it is, as you can see, the file is named `container-build.sh`, it is exactly the script executed by the container.

I guess you noticed that in order to pass environment variables to the script, you need to place them in a file and use `--env` command line option.

**Comment by TommyMurphyTM1234** [comment]
>  I think the overhead would be important, both in terms of build time and archive size.

Yes.

> how does your Windows archive size compare to mine?

The Linux/Windows 64 bit tools archives are c. 350MB.
Yours seem to be about 120-130MB?

> I guess you noticed that in order to pass environment variables to the script, you need to place them in a file and use --env command line option.

Sorry - I didn't. I will have a look at that. Thanks.

**Comment by TommyMurphyTM1234** [comment]
Hi Liviu - sorry to bother you but I'm probably doing something wrong but cannot figure out what.

I have made the original change to container-build.sh to allow GCC_MULTILIB to be overridden externally.

I have and env_file containing the following:

`GCC_MULTILIB=(rv32i-ilp32-- rv32im-ilp32--)`

I have also edited container-gcc-functions-source.sh to echo the argument passed to multilib-generator:

```
        # Be sure the ${GCC_MULTILIB} has no quotes, since it defines 
        # multiple strings.
	echo "${GCC_MULTILIB[@]}"
        ./multilib-generator ${GCC_MULTILIB[@]} >"${GCC_MULTILIB_FILE}"
        cat "${GCC_MULTILIB_FILE}"

```
And I build as follows:

`bash .../scripts/build.sh -env-file env_file --linux64 2>&1 | tee build.log`

But I get this:

```
Running the multilib generator...
(rv32i-ilp32-- rv32im-ilp32--)
Traceback (most recent call last):
  File "./multilib-generator", line 40, in <module>
    (arch, abi, extra, ext) = cfg.split('-')
ValueError: too many values to unpack

```
I have tried different variants of the GCC_MULTILIB variable setting (e.g. with and without the surrounding brackets etc.) but the only time things work is if I specify a single multilib - e.g.:

`GCC_MULTILIB=rv32i-ilp32--`

Anything else fails.

Any ideas and apologies again if this is a dumb question/problem... I suspect that it's simply down to some syntax issue but so far trial and error hasn't helped to identify a working case.

**Comment by TommyMurphyTM1234** [comment]
Actually I tried to pass other parameters from the env file and for some reason the windows 64 bit build failed:

```
Container script "/Host/Work/riscv-none-gcc-7.2.0-4/build.git/scripts/container-build.sh" completed at Tue Jun 19 18:20:10 UTC 2018.
Duration: 3 minutes.
Docker container "riscv-none-gcc-linux64-build" removed.
Mandatory GNU/Linux binaries missing.

```
However the build works when I edit the container-build.sh (to specify my list of multilibs) and defs-source.sh (e.g. to specify the prefix of riscv64-unknown-elf rather than riscv-none-embedded for the tools) and specify the "override" values there instead.

I need to investigate further and provide more specific info as soon as I get a chance but it seems to me that there may be some problems with passing override values via the --env-file mechanism.

**Comment by ilg-ul** [comment]
> GCC_MULTILIB  ... I suspect that it's simply down to some syntax issue but so far trial and error hasn't helped to identify a working case.

I remember having some issues with arrays in bash, but I don't know details :-(

but I guess you are right, it might be some small syntax trick that we are missing.

**Comment by TommyMurphyTM1234** [comment]
Let me look at it again as soon as I get a chance.
Thanks.

**Comment by ilg-ul** [comment]
> there may be some problems with passing override values via the --env-file mechanism

actually I did not use this mechanism, I added it only knowing that you'll probably need it, so, yes, there may be some problems with it.

my suggestion is to check the bash manual for the syntax when overriding arrays.

**Comment by AntonKrug** [comment]
Yes the archive size increases by good bit.

I was thinking it would be because the bash () is the array syntax, while the docker env files are just literal strings? so either it will fail as a array, or it will fail while passing it. maybe it will see it as very long 1 entry instead of multiple smaller ones?

**Comment by AntonKrug** [comment]
I think giving it 1 entry might pass because now whole array == single entry so even when the array handling fails the internal parameter will get single separate arch-abi parsed correctly. So all the time it gets feed all the abis at once and it only works when all arch-abi == single arch-abi

**Comment by ilg-ul** [comment]
did you find a way to pass this from the environment?

**Comment by TommyMurphyTM1234** [comment]
Sorry I never got a chance to revisit this and I continue to use a modified container-build.sh script in order to build a different set of multilibs from the default.

**Comment by TommyMurphyTM1234** [comment]
Fixed by PR #15
